pkg/controller/dnsrecord/actuator.go

```diff
@@ -0,0 +1,142 @@
+	a.client = client
+	return nil
+}
+
+// Reconcile reconciles the DNSRecord.
+func (a *actuator) Reconcile(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
+	// Create AWS client
+	awsClient, err := aws.NewClientFromDNSSecretRef(ctx, a.client, dns.Spec.SecretRef, dns.Spec.Region)
+	if err != nil {
+		return fmt.Errorf("could not create AWS client: %+v", err)
+	}
+
+	// Determine DNS hosted zone ID
+	zone, err := a.getZone(ctx, dns, awsClient)
+	if err != nil {
+		return err
+	}
+
+	// Create or update DNS record
+	ttl := extensionsv1alpha1helper.GetDNSRecordTTL(dns.Spec.TTL)
+	a.logger.Info("Creating or updating DNS record", "zone", zone, "name", dns.Spec.Name, "type", dns.Spec.RecordType, "values", dns.Spec.Values)
+	if err := awsClient.CreateOrUpdateDNSRecord(ctx, zone, dns.Spec.Name, string(dns.Spec.RecordType), dns.Spec.Values, ttl); err != nil {
+		return fmt.Errorf("could not create or update DNS record in zone %s with name %s, type %s, and values %v: %+v", zone, dns.Spec.Name, dns.Spec.RecordType, dns.Spec.Values, err)
+	}
+
+	// Update resource status
+	return extensionscontroller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.client, dns, func() error {
+		dns.Status.Zone = &zone
+		return nil
+	})
+}
+
+// Delete deletes the DNSRecord.
+func (a *actuator) Delete(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
+	// Create AWS client
+	awsClient, err := aws.NewClientFromDNSSecretRef(ctx, a.client, dns.Spec.SecretRef, dns.Spec.Region)
+	if err != nil {
+		return fmt.Errorf("could not create AWS client: %+v", err)
+	}
+
+	// Determine DNS hosted zone ID
+	zone, err := a.getZone(ctx, dns, awsClient)
+	if err != nil {
+		return err
+	}
+
+	// Delete DNS record
+	ttl := extensionsv1alpha1helper.GetDNSRecordTTL(dns.Spec.TTL)
+	a.logger.Info("Deleting DNS record", "zone", zone, "name", dns.Spec.Name, "type", dns.Spec.RecordType, "values", dns.Spec.Values)
+	if err := awsClient.DeleteDNSRecord(ctx, zone, dns.Spec.Name, string(dns.Spec.RecordType), dns.Spec.Values, ttl); err != nil {
+		return fmt.Errorf("could not delete DNS record in zone %s with name %s, type %s, and values %v: %+v", zone, dns.Spec.Name, dns.Spec.RecordType, dns.Spec.Values, err)
+	}
+
+	// Update resource status
+	return nil
+}
+
+// Restore restores the DNSRecord.
+func (a *actuator) Restore(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
+	return a.Reconcile(ctx, dns, cluster)
+}
+
+// Migrate migrates the DNSRecord.
+func (a *actuator) Migrate(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
+	return nil
+}
+
+func (a *actuator) getZone(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, awsClient awsclient.Interface) (string, error) {
+	switch {
+	case dns.Status.Zone != nil && *dns.Status.Zone != "":
```

Comment:
This means, once the zone in the status is set, there is no way to influence it anymore via the spec field.

Just to be on the safe side:
Is there a use case where we might want to prefer the spec over the status field? I could imagine that if you deleted the record and recreated the hosted zone in between reconciliations, the controller would not be able to recover from this even if you specify the new hosted zone in the spec. 
Though, I'm not sure if that's entirely correct, so I'm asking here :)

Reply:
I think this is really minor, but it would be easy to just reverse the 2 cases and make sure the spec takes precedence if it's specified. I will do it.

Reply:
Done.